python_modules/dagster/dagster/_core/definitions/asset_check_result.py

```diff
@@ -53,3 +77,59 @@ def __new__(
             success=check.bool_param(success, "success"),
             metadata=normalized_metadata,
         )
+
+    def to_asset_check_evaluation(
+        self, step_context: "StepExecutionContext"
+    ) -> AssetCheckEvaluation:
+        specs = step_context.job_def.asset_layer.check_specs_for_node(step_context.node_handle)
+
+        spec_check_names_by_asset_key: Dict[AssetKey, Set[str]] = defaultdict(set)
+        for spec in specs:
+            spec_check_names_by_asset_key[spec.asset_key].add(spec.name)
+
+        if self.asset_key is not None:
+            if self.asset_key not in spec_check_names_by_asset_key.keys():
+                raise DagsterInvariantViolationError(
+                    "Received unexpected AssetCheckResult. It targets asset"
+                    f" '{self.asset_key.to_user_string()}' which is not targeted by any of the"
+                    " checks currently being evaluated. Targeted assets:"
+                    f" {[asset_key.to_user_string() for asset_key in spec_check_names_by_asset_key.keys()]}."
+                )
+
+            resolved_asset_key = self.asset_key
+
+        else:
+            if len(spec_check_names_by_asset_key) > 1:
+                raise DagsterInvariantViolationError(
+                    "TODO Check result didn't specify an asset key, but there are multiple assets"
+                    " to choose from: {name the assets}"
+                )
+
+            resolved_asset_key = next(iter(spec_check_names_by_asset_key.keys()))
+
+        spec_check_names_for_asset_key = spec_check_names_by_asset_key[resolved_asset_key]
+        if self.check_name is not None:
+            if self.check_name not in spec_check_names_for_asset_key:
+                raise DagsterInvariantViolationError(
+                    "Received unexpected AssetCheckResult. No checks currently being evaluated"
```

Comment:
i agree with Pedram that these errors will be frustrating to hit, but I also think that if we relax the constraint then we may as well not have check specs

Reply:
Yeah, I think we should keep this at least constraint at least for now and can potentially relax in the future.

Reply:
Much easier to relax than to un-relax.